### ISSUE TITLE
Kerr, Reissner–Nordström, and Kerr-Newman

### ISSUE THREAD

**Issue by WrathfulSpatula** [vm6502q/OpenRelativity#50]
We've had Schwarzschild metric in this fork, for a long time. This is a static, stationary black hole. We should expand the treatment to also encompass rotating and charged black holes.

Hypothetical "new physics" proposed by Strano can be turned off entirely, for Schwarzschild. The same will be true for rotating and charged black hole metrics.

It's been a while since I formally considered how my own work (Strano's) might apply to spinning and charged black holes, but the evaporation of a (spinning) Kerr metric ergosphere happens via the ostensible spin-2 graviton, _not_ a quasi-scalar mode from allowing metric tensor torsion, if I recall. Charged Reissner–Nordström black holes are assumed by the author to emit charge in quasi-scalar packets that are sub-"extremal," and these probably happen to carry multiples of 2 fundamental units of charge in a packet. (This includes the possibility for net 0 charge, in which case they very quickly self-annihilate. The author thinks magnetic binding forces are sufficient to keep these other theoretically "bare" like-sign charges coincident, but, rather, they therefore must carry sufficient energy to avoid being "hyperextremal." in order to exist as "real" particles.)

I'll be working on this, at night over the next few weeks, or however long it takes.

**Comment by WrathfulSpatula** [comment]
To clarify, in (spinning) Kerr metric, the quasi-scalar emission from the interior event horizon undergoes a Penrose process in the ergosphere, separating the quasi-scalar to emit net fundamental spin-2 gravitons. We'll handle this case first, starting tonight.

**Comment by WrathfulSpatula** [comment]
We have an approximation to the Kerr metric, from #51: we treat Kerr as a Schwarzschild space-time whose preferred frame of accelerated rest locally rotates with `omega` parameter calculated as in Wikipedia:

https://en.wikipedia.org/wiki/Kerr_metric#Frame_dragging

According to the linked section of the article, Kerr is strictly a diffeomorphic to Schwarzschild under the local parameterization by this `omega`.

First, we approximated `omega` by the starting point of a finite difference time step, in the `MonoBehavior.Update()` loop. Tonight, we improved the approximation by averaging the starting and ending `omega` value, before and after Schwarzschild metric co-motion for the same time interval. (Better approximations, elliptical or hyperbolic, are possible via spline or integration over the time step.) RelativisticObject instances and the player also intrinsically spin, in co-motion with `omega`.

The Kerr optics still aren't quite right. We might modify our Schwarzschild strong-lensing-at-a-distance blitting shader with a relativistic Doppler shift for tangential velocity from `omega`. For the full-screen blitting shader, which assumes that the black hole is very distant, the effect would be calculated on a plane orthogonal to line-of-sight from player to "world coordinates" origin, but also depending on the angle of this plane with the axis of Kerr spin.

**Comment by WrathfulSpatula** [comment]
More on Kerr lensing: Much has been written about the Kerr lens equations, such as [this article](https://arxiv.org/abs/gr-qc/0308023) and many others, though this topic seems like it might still be a research-level problem, rather than neatly wrapped up in a canonical analytical form.

Not having had time peruse the literature, yet, the article linked above starts with the simpler cases of looking directly at a Kerr black hole on the poles, then equatorial plane, before attempting total generality. We can borrow these same cases, under the assumption that _Kerr metric is diffeomorphic to Schwarzschild under a map by the local `omega` angular velocity parameter_, of its "frame dragging."

Intuitively, under this assumption, looking directly along the axis of rotation of Kerr metric, one expects a Schwarzschild lensing effect diffeomorphically transformed by a "corkscrew" effect from the integration of `omega` over the path between source and observer, under these perfectly aligned conditions. The focal point does not change. Any energy difference in the photons is cyclically cancelled as with Schwarzschild metric, between a distant source and a distant observer.

First, I will handle the case of integrating this `omega` parameter as a diffeomorphism on Schwarzschild, causing this "corkscrew" effect. Then, I will consider the equatorial plane case, including Doppler shift. Then, I'll try to put together an approximation for rotating the observer's viewpoint between these two extremes. (The third spatial dimension to consider is only the rotation of the on-face plane, for the old Schwarzschild lens shader, which is comparatively "trivial.")

**Comment by WrathfulSpatula** [comment]
In the case of the equatorial plane, approximating by the locally varying `omega` parameter on Schwarzschild, the effect is probably to shift the source point of the focus toward the away-spinning side (relative the observer) of the black hole, dependent upon black hole spin and inclination angle. Additionally, there might be Doppler blue-shift on the toward-spinning side, (toward the observer,) and red-shift on the away-spinning side.

We'll form the general approximation basically by spherically interpolating between these on-pole and on-equator cases, while, again, the third spatial dimension is just (Euclidean) rotation of the on-face blit plane of the shader.

**Comment by WrathfulSpatula** [comment]
We've implemented this, as described above. Let's have a word of recap.

Gravitational lensing by Schwarzschild is commonly simplified in approximation to the assumption of a point-like deflector in front of a very distant lensed background, with the observer at a specified finite distance.

If the angle of deflection is `4GM/(rc^2)`, then this is the same as `-4/c^2` times the gravitational potential. Thinking of the deflection as due to interaction with a conservative gravitational mass field potential, we expect equivalent interaction with a conservative gravitational angular momentum potential, directly proportional to fraction of angular momentum over Schwarzschild radius, in Planck units.

Classical intuition tells us this angular momentum potential interaction is entirely and exactly polarizing, looking directly along the Kerr axis of rotation; the equatorial term implies the same conservative angular momentum potential interaction for deflections exactly planar to the equator, dropping off as `cos(inclination)`.

Hopefully, it's a reasonable theoretical approximation.

**Comment by WrathfulSpatula** [comment]
I've been wrestling since last night with whether the maximum deflection in the equatorial plane is exactly the same as along the axis, or whether different by some factor like 2 or 4. I think I have a better argument, that these two are the same.

Consider 2 path integrals of `omega` approximating "straight lines" in the coordinates, one along the axis of Kerr spin, and the other in the equatorial plane. The maxima of the integrand functions should exactly match for these two, since they are effectively the same point, for `\omega`. The scaling factor between the two integrated areas would be the scaling between spin-axis-orientation and equatorial-orientation.

Along the spin axis path, the value of `\omega` follows a half period of `cos(inclination)`. If the integrand takes a maximum value of 1, then the area of the integral is 2. `\omega` is _always_ entirely at a right angle to the "velocity" of a path along the spin axis.

Integrating a path straight through the (assumed point-like) deflector, in the _equatorial_ plane, the path integral's "velocity" is no longer at a right angle to `\omega`; instead, it sinusoidally varies, as with the spin axis path. The perpendicular component is `sin(azimuth)`, and the parallel component is therefore `cos(azimuth)`. The _parallel_ component is frame-dragging all trajectories within the equatorial plane.

By heurism, `cos(azimuth)` and `cos(inclination)` have the same definite area over the same domain _if_ they have the same amplitude scale. Hence, I _think_ the amplitudes scale is 1:1, as is already implemented in the shader.

**Comment by WrathfulSpatula** [comment]
Note that there is no new Doppler shift effect for Kerr, beyond Schwarzschild, since this is a function of the _energy change_ of photons being lensed, which is _strictly_ a function of the conservative _mass potential_ field, not the conservative _angular momentum potential_ field.

**Comment by WrathfulSpatula** [comment]
By the way, we find a happy, self-consistent result, for a conservative angular momentum potential field, when we consider arbitrary rotation of the spin axis: no matter what angles the observer's position forms with the spin axis and equator, the _magnitude_ of deflection by the the gravitational effect of angular momentum is _exactly the same_, and depends only on the distance of the observer from the center of the metric. However, the _partitioning_ into orthogonal _polarizing_ and _non-polarizing_ terms depends upon _inclination angle_.

**Comment by WrathfulSpatula** [comment]
With Kerr metric sufficiently handled for now, I'm thinking about how to address **Reissner-Nordström**, for a charged black hole.

Sorry for the 20-year-old Geocities-type page link, (from University of Colorado, Boulder,) but, then, **Reissner-Nordström** has been known for nearly the full 100+ years since the inception of the theory of general relativity: 

https://jila.colorado.edu/~ajsh/bh/rn.html

What's made Kerr metric so easy to handle has been its diffeomorphic mapping to Schwarzschild metric via the locally varying `\omega` parameter, for angular velocity of frame-dragging. With an implementation of the Schwarzschild `ConformalMap` type in hand, and the `\omega` parameter, we can treat Kerr metric as just a very terse transformation layer basically laid _on top_ of Schwarzschild, (or endpoint-averaged over a finite difference time step, as has ultimately become our approach). To make any `ConformalMap` work, we need **free fall coordinates** for `ComoveOptical()`, and it'd be great if we could _directly_ use Schwarzschild as a base layer for this comovement transformation, again, as with Kerr.

The link above points out that Reissner-Nordström free fall coordinates take the same form as Schwarzschild, though, if we think of the black hole's _mass_ as _locally varying over the radial coordinate_. (`M(r) = M + Q^2/(2*r)`)

Then, if we modify our `Schwarzschild` object to require a _"position-in-world,"_ "`piw`", in order to determine the _Schwarzschild radius_, then we can just treat the free fall coordinates virtually as if they were Schwarzschild!

I might have a working implementation tonight, on this basis.

**Comment by WrathfulSpatula** [comment]
With the last couple of commits, Reissner-Nordström turns out to be even _much_ easier than Kerr.

Every `RelativisticObject` acts _locally_ as if Reissner-Nordström were exactly Schwarschild, with an effective Schwarzschild radius parameterized on the radial coordinate. (For a finite difference time step, we lose some comovement accuracy due to the effective radius _changing continuously over the course of the time step_, though we could take multiple partial samples and compose the comovement over the proper time interval to alleviate this, but we don't necessarily need to for "video game accuracy" purposes.) This includes the _player perspective_ effective radius, which determines how the Schwarzschild radius _looks_ for the _shader_.

**Comment by WrathfulSpatula** [comment]
Our speculative black hole evaporation, in this case, implies that the gravitational field can carry _charge_. (Might as well call these particles "**gravitrons**.") It has to, for this sort of evaporation to work, and this is where we get _really_ far out, from accepted canonical theory, admittedly.

By Fermi's golden rule, due to the implied extra particle species' Feynman diagrams, we have an additional (overall charge-neutral) evaporation term, where pairs of "gravitrons" carrying net neutral charge immediately self-annihilate to produce a thermal gas of photons, a true luminous "glow." This "should" technically be there in Schwarzschild and Kerr, as well, (if we follow the author's hypothesis,) but we'll reserve it for Reissner-Nordström and Kerr-Newman, (both being charged black hole metrics). At least, this is a reasonable "separation of concerns," from a software development perspective.

(The only other note I'll offer, on this particular theoretical point, is that this additional "gravitron" field mode term, evaporating into photons, _should_ be visible in the thermal spectrum of _all_ **stellar** black hole candidates we know, whereas active galactic nuclei are too large to exhibit it with any significant intensity. Of all the candidates we have observed, for black holes formed from stellar collapse, these candidates tend to be in binary systems, with a partner, and we have microwave spectra observation data for most of them. While the preferred theory for "microwave excess" in these spectra is synchrotron radiation, if there even is an "excess," per se, Wien's displacement law on the back of a literal envelop doesn't give a result entirely beyond the realm of plausibility, for my favorite pet competing explanation.)

**Comment by WrathfulSpatula** [comment]
We're going to combine the mass, spin, and charge pieces we already have into Kerr-Newman metric first, though, and worry about that later in the weekend. The previous comment is a matter of personal satisfaction with the already non-canonical `doEvaporate` checkbox behavior.

**Comment by WrathfulSpatula** [comment]
That's done, but I realize that I missed the effective displacement of the Schwarzschild-like (inner) event horizon for Kerr. Luckily, this is easily handled like Reissner-Nordström.

**Comment by WrathfulSpatula** [comment]
I misread this section of the Wikipedia article on Kerr metric:

https://en.wikipedia.org/wiki/Kerr_metric#Frame_dragging

The diffeomorphic relationship to Schwarzschild is directly via components of the metric tensor, which I have _not_ used as the effective Schwarzschild radius. While replacing the effective radius with the appropriate term from these coordinates will serve for comovement, it might not match the expectation for optics, as the gravity lens shader calculates.

I will debug.

**Comment by WrathfulSpatula** [comment]
Sorry for the commit flailing, but there is incremental progress. Kerr is much more difficult to analyze, compared to Reissner-Nordström.

For Reissner-Nordström metric, it's nearly "trivial" to treat it as Schwarzschild metric with a locally varying Schwarzschild radius parameterized on the radial coordinate. Optically, the player sees the Schwarzschild radius parameterization value for the player's own immediate "world (3-)position." At any point in the metric, on a given time coordinate hypersurface, we use the appropriate Schwarzschild radius parameterization value for the `RelativisticObject` "world 3-position."

We can't just immediately extend this kind of parameterization "trick" to Kerr, maybe at all. Also, of all the coordinates we could define on the Kerr metric, even if the mechanical paths are diffeomorphic to the correct geodesics, we might _not_ have the right coordinates to show the player how the mechanics would "really" _look_, from the player's perspective.

But, unflagging, we go ahead and try to adapt the parameterized effective Schwarzschild radius approach from Reissner-Nordström, but instead for Kerr. Arbitrarily, say that we rescale just by the respective time coordinate differential factors, between Kerr and Schwarzschild: `1-(r_s)*r/sigma = 1-(r_s')/r`. `sigma` depends on the inclination relative the Kerr equator; this parameterization of an effective Schwarzschild radius is equal to _the invariant mass Schwarzschild radius_ directly along the axis of rotation. So far, it's plausible that, at exactly at this axis of perfect symmetry, an effective Schwarzschild radius appears the same as the _invariant mass_ Schwarzschild radius. Further, our effective radius parameterization responds to the inclination.

Straight from the Kerr metric Wikipedia article, we have an expression for the scale of time-like coordinate for an observer at infinity relative the first-person proper time of an object closer to the spinning black hole:

![efc4704dc10943543568e6a13fb7217b20a4f405](https://user-images.githubusercontent.com/2674693/128614952-9828589f-7f85-4055-bd5a-c34db6e1eb31.png)

This is analogous to the familiar `t'/tau=1/(1-r_s/r)^(1/2)` for Schwarzschild metric, which we've used (and abused) to approximate the behavior of Kerr up to the `omega` angular velocity Killing horizon. If we put one expression over the other, to form `t/t'`, we get a coordinate scaling factor between time-like coordinates for an observer at infinity for Kerr, over that for an observer at infinity for Schwarzschild. Note that this time scale is _not_ generally 1:1 at the Kerr poles, even though our effective Schwarzschild radius is! This makes a certain amount of sense, though, as similarly to our discussion of conservative _mass potential_ and _angular momentum potential_ fields for the optical lensing of Kerr went, earlier, our "mass potential" effect is augmented by a nonzero "angular momentum potential" whose _magnitude is independent of observer angle_, we found for lensing; there's intrinsic _magnitude_ of _angular momentum gravity_ depending only on _radius_.

So, we endpoint-average the Killing horizon angular momentum on top of this to finish our comovement and local effective acceleration calculations, right? I guess, but I need better tests. There's a certain amount we can get away with, in numerical finite difference simulation, as opposed to the symbolic math, but I need some distance from this to form a plan for some kind of acceptable validation, even if just via symbolic derivation.
